MetaMask/metamask-extension
# feat(restore): Restore vault when state corruption occurs

## **Description**
This pull request adds an automatic saving of the encrypted keyring vault to localStorage, as well as the last good migration number seen. We persist the lastGoodMigration key after migrator runs so it should always be up to date. We compare the value to the persisted state before running the migrator as another way of detecting a corrupted state tree. 

We can, in the case of corruption, load the vault from localStorage and update the initialStateTree with that value (as well as onboarding completed to avoid a routing loop) and the extension can recover. 

CONS:
any third party optin preferences, metametrics, etc will be wiped and the onboarding flow has been bypassed. there's probably a cleaner way to avoid the loop but without this I just keep getting routed back to a full screen onboarding-type flow with an unlock screen. Typing password and submitting just lands me back on that page. 

**edit**
At @danjm request/suggestion I added an error handler for this state that will show the MetaMask error screen when the state restore happens. Screeenshots included below. 

## **Related issues**

Fixes: 

## **Manual testing steps**
1. Run `yarn start`
2. Load the extension
3. Unlock your wallet
4. On the background process, inspect and go to application tab.
5. Inspect local storage.
6. See you have a 'lastGoodMigration' and 'metaMaskVault' key saved. 
7. In the console run `chrome.storage.local.set({ data: null, meta: null });`
8. Restart the extension
9. Load the extension. 
10. See the error message in the screenshots section. 
11. Click restart
12. Load extension
13. See your accounts are set, but you'll get all notifications and etc as a new user would. 

## **Screenshots/Recordings**
<img width="367" alt="Screenshot 2024-02-26 at 3 44 33 PM" src="https://github.com/MetaMask/metamask-extension/assets/4448075/142025cf-8219-41c3-b96b-1a49bbfe775a">


### **Before**



### **After**



## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
